Implement notched box plots, closes #2286

krassowski opened the pull request:
I added a boolean option "notched" (default false) and a numerical "notchwidth" (default 0.5).
The options define if the box trace should be notched or not, and if yes, determine the width of the notch (proportional to the width of the box).

I modified the svg path-generation code, utilizing conditional (ternary) operators so the common parts of the box (top & bottom edges for vertical, left & right edges for horizontal) are shared.

I added two test baselines: `box_notched` and `box_horz_notched`.

![box notched](https://raw.githubusercontent.com/krassowski/plotly.js/a234668c7af2e2409052fe06817f939b931dad5f/test/image/baselines/box_notched.png)

![box horiz notched](https://raw.githubusercontent.com/krassowski/plotly.js/144989c7e97aad5ae3a1a6ab4a48eb642024e49b/test/image/baselines/box_horz_notched.png)

krassowski committed:
Implement notched box plots, closes #2286

alexcjohnson commented on src/traces/box/defaults.js:
```diff
@@ -29,6 +29,9 @@ function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
     coerce('boxmean');
 
+    coerce('notched');
+    coerce('notchwidth');
```
We should only coerce `notchwidth` if `notched` is `true`.

alexcjohnson commented on src/traces/box/attributes.js:
```diff
@@ -85,6 +85,28 @@ module.exports = {
+        description: [
+            'Sets the width of the notches relative to',
+            'the box\' width.',
+            'For example, with 1, the whiskers are as wide as the box(es).'
```
It looks like Matlab and ggplot have opposite conventions about this... if I'm reading it right this is the ggplot version. I would have probably gone for the Matlab version, where `notchwidth=0` is nothing taken away, and `notchwidth=0.5` is notches that meet in the middle (actually I might have called that `1` instead... but we don't want to introduce yet ANOTHER convention). I guess I don't have that strong an opinion about it though, @etpinard or @cpsievert do you care which we use?

krassowski replied:
Good point; I will wait for others to express their opinion. Just to have it here: the behavior (and description) of `notchwidth` is based on another, already existing attribute - `whiskerwidth`.

alexcjohnson commented on test/image/mocks/box_horz_notched.json:
```diff
@@ -0,0 +1,287 @@
+{
+   "data":[
```
Very nice test images! re: the labels getting cut off - that'll get fixed by https://github.com/plotly/plotly.js/pull/2243 but here we might as well just give it a larger left margin so they fit.

alexcjohnson commented on src/traces/box/plot.js:
```diff
@@ -135,18 +139,30 @@ function plotBoxAndWhiskers(sel, axes, trace, t) {
-                    'M' + q1 + ',' + pos0 + 'V' + pos1 + 'H' + q3 + 'V' + pos0 + 'Z' + // box
+                    'M' + m + ',' + posm0 + 'V' + posm1 + // median line
+                    'M' + q1 + ',' + pos0 + 'V' + pos1 + // left edge
+                    (((notched === true)) ? 'H' + ln + 'L' + m + ',' + posm1 + 'L' + un + ',' + pos1 : '') + // top notched edge
```
That's a lot of `(())`... I think it can be just `(notched ? ... : '')`

alexcjohnson reviewed (commented):
@krassowski this is great, thanks for adding the feature! I just made a few comments, feel free to PR to the main repo after fixing.  We may have to wait until Monday to see if anyone else feels strongly about the `notchwidth` convention.

krassowski committed:
Address 3/4 of @alexcjohnson's comments on notches

dy committed:
Unbind VAO in gl-surface3d

dy committed:
Bump gl-surface3d

dy committed:
Update mocks

dy committed:
Remove gl-shader direct dep

dy committed:
Merge pull request #2306 from plotly/unlock-shader-fix

Unbind VAO in gl-surface3d

krassowski committed:
Merge branch 'master' into notched_box_plot

krassowski committed:
Implement 'smarter' notched/notchwidth default

I deliberately removed `"notched":true` in one of test
box traces so it is (indirectly) tested

krassowski committed:
Change notchwidth convention to Matlab, d=0.25

krassowski committed:
Set notch width range to (0, 0.5) instead of (0,1)

presumably for better consistency with Matlab

krassowski committed:
Update test images for notched box plots
